### PR TITLE
[NFC] First activation solver

### PR DESCRIPTION
**Pull request by DrizztDoUrden**
This implements the first solver for activation. For now, I have taken the smallest possible PR approach, so the activation forward is split to the part with the solver and the rest, which would be translated to another solver and removed in the next PR.
I have kind of no idea how to name solvers here and would gladly accept the suggestions.
This PR also contains some essentials like activation problem description, and activation invoke parameters, which may change in the future.

Performance has been compared on a synthetic test of 105 activation forward configs with the following results: +0.91% avg GPU performance, +0.46% min GPU performance, +8.45% wall clock performance. I would like to note that this is probably not meaningful. For some reason, time is extremely volatile. Even when I run develop a couple of times in a row it varies significantly for small configs. I would experiment with larger configs and would attempt to come up with better test set.

### CODE REVIEW

**Comment by DrizztDoUrden** [comment]
Status update: this has some performance regressions according to my tests. I am currently debugging the PR and the test.

**Comment by atamazov** [comment]
@DrizztDoUrden 
> Performance has been compared...

The GPU performance should remain the same provided that this PR does not have bugs.

What is interesting is how this change affects the host-only overhead (shown by the driver for convolutions as "Aux Wall Time")

**Comment by DrizztDoUrden** [comment]
I have 2.25% wall clock, 0.71% - gpu min and 0.73% - gpu avg time difference when running develop twice in a row.

**Comment by atamazov** [comment]
I am going to merge this as soon as CI tests passed.

File: src/include/miopen/find_solution.hpp [hunk 1]
```diff
@@ -182,6 +182,55 @@ struct SolverContainer
...
+                    }
+                    else
+                    {
+                        /// \todo If Solver is applicable it must provide an appropriate Solution.
+                        /// This is not the case for some 20x5 convolutions (and possibly others).
+                        /// Normally we should not get here and message level should be Error.
+                        /// For now, let's use Info (not Warning) level to avoid
+                        /// flooding the console.
+                        MIOPEN_LOG_I(SolverDbId(solver)
+                                     << ": [Warning] Applicable Solver not succeeded.");
```
Comment (atamazov) [inline src/include/miopen/find_solution.hpp:226]: Do it in your next PR:
```suggestion
                        MIOPEN_LOG_E(SolverDbId(solver)
                                     << ": Applicable Solver not succeeded.");
```

File: src/include/miopen/find_solution.hpp [hunk 2]
```diff
@@ -182,6 +182,55 @@ struct SolverContainer
...
+            [&](auto solver) {
+                if(count >= limit)
+                    return;
+                if(find_only.IsValid() && find_only != Id{SolverDbId(solver)})
+                { // Do nothing (and keep silence for the sake of Tuna), just skip.
+                }
+                // For better performance, check IsDynamic() first, because
+                // it is much faster than IsApplicable().
+                // else if(problem.use_dynamic_solutions_only && !solver.IsDynamic())
+                //    MIOPEN_LOG_I2(SolverDbId(solver) << ": Skipped (non-dynamic)");
```
Comment (atamazov) [inline src/include/miopen/find_solution.hpp:205]: [future] I think we need to keep the notion of "dynamic" solutions for all primitives.

File: src/solver.cpp [hunk 3]
```diff
@@ -417,6 +425,8 @@ inline SolverRegistrar::SolverRegistrar(IdRegistryData& registry)
     RegisterWithSolver(registry, ++id, ConvMlirIgemmFwdXdlops{}, miopenConvolutionAlgoImplicitGEMM);
     RegisterWithSolver(registry, ++id, ConvMlirIgemmBwdXdlops{}, miopenConvolutionAlgoImplicitGEMM);
     RegisterWithSolver(registry, ++id, ConvMlirIgemmWrWXdlops{}, miopenConvolutionAlgoImplicitGEMM);
+
+    Register(registry, ++id, SolverDbId(activ::ActivFwdSolver0{}));
```
Comment (atamazov) [inline src/solver.cpp:429]: If we have single registry for all primitives, won't this affect performance?

File: src/solver/activ/fwd_0.cpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+                                                                   ? x_strides[1]
+                                                                   : (x_lens_.size() == 4)
+                                                                         ? x_strides[2]
+                                                                         : x_strides[3]);
+                    const auto y_stride2D = static_cast<unsigned int>(
+                        (y_lens_.size() == 2) ? y_strides[0] : (y_lens_.size() == 3)
+                                                                   ? y_strides[1]
+                                                                   : (y_lens_.size() == 4)
+                                                                         ? y_strides[2]
+                                                                         : y_strides[3]);
```
Comment (atamazov) [inline src/solver/activ/fwd_0.cpp:194]: It seems like many computations are invariant here and can be moved out from the Invoker and even out from the Factory. Shall we open a ticket for this optimization?

**Review by atamazov — APPROVED** [review]
Good to go!

File: src/solver.cpp (hunk 3 above)
Comment (DrizztDoUrden) [reply src/solver.cpp:429]: One would need hundreds of thousands of solvers to nothing something even on synthetic tests.

File: src/solver/activ/fwd_0.cpp (hunk 4 above)
Comment (DrizztDoUrden) [reply src/solver/activ/fwd_0.cpp:194]: Oh, but they are not invariant. You are used to convolution, I have been too ;) Tensor descriptors are not part of network config in activation.

Comment (atamazov) [reply src/solver/activ/fwd_0.cpp:194]: Wow.

Comment (DrizztDoUrden) [reply src/solver/activ/fwd_0.cpp:194]: Took me a while to even consider that possibility.

File: src/include/miopen/find_solution.hpp (hunk 1 above)
Comment (atamazov) [reply src/include/miopen/find_solution.hpp:226]: @DrizztDoUrden agreed?

File: src/include/miopen/find_solution.hpp (hunk 2 above)
Comment (atamazov) [reply src/include/miopen/find_solution.hpp:205]: @DrizztDoUrden ok?

Comment (DrizztDoUrden) [reply src/include/miopen/find_solution.hpp:205]: Yeah, probably

File: src/include/miopen/find_solution.hpp (hunk 1 above)
Comment (DrizztDoUrden) [reply src/include/miopen/find_solution.hpp:226]: ok

**Pull request merged**